compiler/rustc_builtin_macros/src/deriving/default.rs

```diff
@@ -92,11 +104,47 @@ fn default_enum_substructure(
         default_variant
     } {
         Ok(default_variant) => {
-            // We now know there is exactly one unit variant with exactly one `#[default]` attribute.
-            cx.expr_path(cx.path(
-                default_variant.span,
-                vec![Ident::new(kw::SelfUpper, default_variant.span), default_variant.ident],
-            ))
+            // We now know there is exactly one variant with exactly one `#[default]` attribute.
+            match &default_variant.data {
+                VariantData::Unit(_) => cx.expr_path(cx.path(
+                    default_variant.span,
+                    vec![Ident::new(kw::SelfUpper, default_variant.span), default_variant.ident],
+                )),
+                VariantData::Struct { fields, .. } => {
+                    // This only happens if `#![feature(default_field_values)]`. We have validated
+                    // all fields have default values in the definition.
+                    let default_fields = fields
+                        .iter()
+                        .map(|field| {
+                            cx.field_imm(
+                                field.span,
+                                field.ident.unwrap(),
+                                match &field.default {
+                                    // We use `Default::default()`.
+                                    None => default_call(cx, field.span),
```

Comment:
Given this, I filed https://github.com/rust-lang/rfcs/pull/3683 as this bit is independent of the default values RFC.